### PR TITLE
Document X509_check_private_key and add test cases

### PR DESCRIPTION
**Pull request by InfoHunter**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

Document X509_check_private_key and X509_REQ_check_private_key, and also test cases for them are added.

Besides, I find X509_check_private_key function doesn't check the input 'EVP_PKEY' is a private key or not. So I put a 'BUGS' section is the pod file. I don't know if it's designed on purpose, but I have seen case that ran into error, in which an EVP_PKEY contained a public key was passed into SSL_use_private_key (which uses X509_check_private_key to check pkeys) and got no error...

### CODE REVIEW

**Comment by InfoHunter** [comment]
Could someone please have a glance at this change?

File: test/x509_check_cert_pkey_test.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    } else if (strcmp(e, "failed") == 0) {
+        expected = 2;
+    } else {
+        TEST_error("invalid 'expected'");
+        goto failed;
+    }
+
+    /* process private key */
+    bio = BIO_new_file(k, "r");
+    if (bio == NULL) {
```
Comment (mattcaswell) [inline test/x509_check_cert_pkey_test.c]: A style more consistent with the rest of the tests would be:

    if (!TEST_ptr(bio = BIO_new_file(k, "r")))
        goto failed

Comment (mattcaswell) [inline test/x509_check_cert_pkey_test.c]: Other similar occurrences of this below.

File: test/x509_check_cert_pkey_test.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    /* process cert or cert request, use the same local var */
+    bio = BIO_new_file(c, "r");
+    if (bio == NULL) {
+        TEST_error("create BIO for cert or cert req failed");
+        goto failed;
+    }
+
+    switch (type) {
+        case 1:
```
Comment (mattcaswell) [inline test/x509_check_cert_pkey_test.c]: Line up case with switch

File: test/x509_check_cert_pkey_test.c [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+out:
+    if (bio)
+        BIO_free(bio);
+    if (x509)
+        X509_free(x509);
+    if (x509_req)
+        X509_REQ_free(x509_req);
+    if (pkey)
+        EVP_PKEY_free(pkey);
```
Comment (mattcaswell) [inline test/x509_check_cert_pkey_test.c]: All of these free routines can take a NULL input, so remove the "if"s above.

File: test/x509_check_cert_pkey_test.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    return ret;
+
+failed:
+    ret = 1;
+    goto out;
```
Comment (mattcaswell) [inline test/x509_check_cert_pkey_test.c]: A more normal style would be for the function to return 0 on failure and 1 on success, and clean up like this:

        ret = 1;
     failed:
        BIO_free(bio);
        X509_free(x509);
        X509_REQ_free(x509_req);
        EVP_PKEY_free(pkey);
        return ret;

**Comment by InfoHunter** [comment]
Yup, coding style has been fixed based on comments. Ping @mattcaswell

**Comment by richsalz** [comment]
The manpage could be ported to 1.0.2 if so inclined ...

**Comment by richsalz** [comment]
clicked on wrong button.

**Comment by mattcaswell** [comment]
Pushed.

**Pull request closed without merging**